src/runtime_src/core/tools/xbutil2/OO_MemRead.cpp

```diff
@@ -105,38 +111,63 @@ OO_MemRead::execute(const SubCmdOptions& _options) const
     throw xrt_core::error(std::errc::operation_canceled);
   }
 
+  if (m_count <= 0) {
+    std::cerr << "ERROR: Please specify a number of blocks greater than zero\n";
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+
   try {
     //-- base address
     addr = std::stoll(m_baseAddress, nullptr, 0);
   }
-  catch(const std::invalid_argument&) {
+  catch (const std::invalid_argument&) {
     std::cerr << boost::format("ERROR: '%s' is an invalid argument for '--address'\n") % m_baseAddress;
     throw xrt_core::error(std::errc::operation_canceled);
   }
 
+  // Validate the number of bytes to be written if defined
+  // This does not need to be defined for the --input option path
+  uint64_t size = 0;
   try {
-    //-- size
-    size = std::stoll(m_sizeBytes, nullptr, 0);
+    if (!m_sizeBytes.empty()) {
+      size = XBUtilities::string_to_base_units(m_sizeBytes, XBUtilities::unit::bytes);
+    }
   }
-  catch(const std::invalid_argument&) {
-    std::cerr << boost::format("ERROR: '%s' is an invalid argument for '--size'\n") % m_sizeBytes;
+  catch (const xrt_core::error& e) {
+    std::cerr << boost::format("Value supplied to --size is invalid: %s\n") % e.what();
     throw xrt_core::error(std::errc::operation_canceled);
   }
 
   XBU::verbose(boost::str(boost::format("Device: %s") % xrt_core::query::pcie_bdf::to_string(xrt_core::device_query<xrt_core::query::pcie_bdf>(device))));
   XBU::verbose(boost::str(boost::format("Address: %s") % addr));
-  XBU::verbose(boost::str(boost::format("Size: %s") % size));
+  XBU::verbose(boost::str(boost::format("Size: %llu") % size));
+  XBU::verbose(boost::str(boost::format("Block count: %d") % m_count));
   XBU::verbose(boost::str(boost::format("Output file: %s") % m_outputFile));
+  XBU::verbose(boost::str(boost::format("Bytes to read: %lld") % (m_count * size)));
 
   //read mem
   XBU::xclbin_lock xclbin_lock(device);
-  
+
+  // Open the output file and write the data as we receive it
+  std::ofstream out_file(m_outputFile, std::ofstream::out | std::ofstream::binary | std::ofstream::app);
   try{
-    xrt_core::mem_read(device.get(), addr, size, m_outputFile);
-  } catch(const xrt_core::error& e) {
+    for(int c = 0; c < m_count; c++) {
```

Comment:
Never did like one character names.  Could you rename 'c' to something like "runningCount" and make it an unsigned integer.